File: day5/src/main.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;
use itertools::Itertools;
use std::cmp;

fn part1(filename: &str) -> io::Result<()> {
    let path = Path::new(filename);
    let file = File::open(&path)?;
    let reader = io::BufReader::new(file);

    let mut seeds: Vec<i64> = vec![];
    let mut new_seeds: Vec<i64> = vec![];
    let mut used: Vec<i64> = vec![];

    for line in reader.lines() {
        let line = line.unwrap();
        if line == "" {
            continue;
        }
        if seeds.len() == 0 {
            let (_, sds) = line.split_once(": ").unwrap();
            for sd in sds.split(" ") {
                let seed = sd.parse::<i64>().unwrap();
                seeds.push(seed);
            }
        }
        else if line.contains(":") {
            if new_seeds.len() > 0 {
                for i in 0..seeds.len() {
                    let el_i = i as i64;
                    if !used.contains(&el_i) {
                        new_seeds.push(seeds[i]);
                    }
                }
                seeds = new_seeds;
            }
            new_seeds = vec![];
            used = vec![];
        } else {
            let parts: Vec<&str> = line.split_whitespace().collect();
            let (destination, source, range) = match &parts[..] {
                [destination, source, range] => (destination, source, range),
                _ => panic!("Unexpected number of parts"), // or handle error differently
            };
            let destination = destination.parse::<i64>().unwrap();
            let source = source.parse::<i64>().unwrap();
            let range = range.parse::<i64>().unwrap();

            for i in 0..seeds.len() {
                let el_i = i as i64;
                if used.contains(&el_i) {
                    continue;
                }
                let seed = seeds[i];
                if seed >= source && seed <= source + range {
                    used.push(i as i64);
                    new_seeds.push(destination + (seed - source));
                }
            }
        }
    }

    println!("Part 1: {}", seeds.iter().min().unwrap());

    Ok(())
}

fn part2(filename: &str) -> io::Result<()> {
    let path = Path::new(filename);
    let file = File::open(&path)?;
    let reader = io::BufReader::new(file);

    let mut seeds: Vec<(i64, i64)> = vec![];
    let mut loaded_seeds = false;
    let mut new_seeds: Vec<(i64, i64)> = vec![];

    for line in reader.lines() {
        let line = line.unwrap();
        if line == "" {
            continue;
        }
        if !loaded_seeds {
            let (_, sds) = line.split_once(": ").unwrap();
            for (sd, range) in sds.split(" ").tuples() {
                let seed = sd.parse::<i64>().unwrap();
                let range = range.parse::<i64>().unwrap();

                seeds.push((seed, seed + range - 1));
            }
            loaded_seeds = true;
        }
        else if line.contains(":") {
            for new_seed in new_seeds {
                seeds.push(new_seed);
            }
            new_seeds = vec![];
        } else {
            let parts: Vec<&str> = line.split_whitespace().collect();
            let (destination, source, range) = match &parts[..] {
                [destination, source, range] => (destination, source, range),
                _ => panic!("Unexpected number of parts"), // or handle error differently
            };
            let destination = destination.parse::<i64>().unwrap();
            let source = source.parse::<i64>().unwrap();
            let range = range.parse::<i64>().unwrap();

            let new_interval = (source, source + range - 1);

            for i in (0..seeds.len()).rev() {
                let seed = seeds[i];
                if seed.0 >= new_interval.0 && seed.1 <= new_interval.1 {
                    // |--------| - mapping interval
                    //   |----|   - seed interval
                    new_seeds.push((destination + (seed.0 - new_interval.0), destination + (seed.1 - new_interval.0)));
                    seeds.remove(i);
                } else if new_interval.0 >= seed.0 && new_interval.1 <= seed.1 {
                    //   |----|   - mapping interval
                    // |--------| - seed interval
                    new_seeds.push((destination, destination + (range - 1)));
                    seeds.remove(i);
                    seeds.push((seed.0, new_interval.0 - 1));
                    seeds.push((new_interval.1 + 1, seed.1));
                } else if seed.0 >= new_interval.0 && seed.0 <= new_interval.1 {
                    // |----|     - mapping interval
                    //   |----|   - seed interval
                    new_seeds.push((destination + (seed.0 - new_interval.0), destination + (new_interval.1 - new_interval.0)));
                    seeds.remove(i);
                    seeds.push((new_interval.1 + 1, seed.1));
                } else if seed.1 >= new_interval.0 && seed.1 <= new_interval.1 {
                    //     |----| - mapping interval
                    //   |----|   - seed interval
                    new_seeds.push((destination, destination + (seed.1 - new_interval.0)));
                    seeds.remove(i);
                    seeds.push((seed.0, new_interval.0 - 1));
                }
            }
        }
    }

    let mut ans = i64::MAX;
    for seed in seeds {
        ans = cmp::min(ans, seed.0);
    }

    println!("Part 2: {}", ans);

    Ok(())
}

fn main() {
    part1("./src/input.txt").unwrap();
    part2("./src/input.txt").unwrap();
}

File: day5/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type segment struct {
	x1, y1, x2, y2 int
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func sign(a int) int {
	if a < 0 {
		return -1
	} else if a > 0 {
		return 1
	}
	return 0
}

func part1(filename string) {
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	ans := 0

	var matrix [1000][1000]int

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		s := strings.Split(scanner.Text(), " -> ")
		segment := segment{}
		fmt.Sscanf(s[0], "%d,%d", &segment.x1, &segment.y1)
		fmt.Sscanf(s[1], "%d,%d", &segment.x2, &segment.y2)

		if segment.y1 == segment.y2 {
			for i := min(segment.x1, segment.x2); i <= max(segment.x1, segment.x2); i++ {
				matrix[segment.y1][i]++
			}
		} else if segment.x1 == segment.x2 {
			for i := min(segment.y1, segment.y2); i <= max(segment.y1, segment.y2); i++ {
				matrix[i][segment.x1]++
			}
		}
	}

	for i := 0; i < 1000; i++ {
		for j := 0; j < 1000; j++ {
			if matrix[i][j] >= 2 {
				ans++
			}
		}
	}

	fmt.Println("Part 1:", ans)
}

func part2(filename string) {
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	ans := 0

	var matrix [1000][1000]int

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		s := strings.Split(scanner.Text(), " -> ")
		segment := segment{}
		fmt.Sscanf(s[0], "%d,%d", &segment.x1, &segment.y1)
		fmt.Sscanf(s[1], "%d,%d", &segment.x2, &segment.y2)

		if segment.y1 == segment.y2 {
			for j := min(segment.x1, segment.x2); j <= max(segment.x1, segment.x2); j++ {
				matrix[segment.y1][j]++
			}
		} else if segment.x1 == segment.x2 {
			for i := min(segment.y1, segment.y2); i <= max(segment.y1, segment.y2); i++ {
				matrix[i][segment.x1]++
			}
		} else {
			j, i := segment.x1, segment.y1
			for {
				matrix[i][j]++

				j += sign(segment.x2 - segment.x1)
				i += sign(segment.y2 - segment.y1)
				if j == segment.x2 && i == segment.y2 {
					matrix[i][j]++
					break
				}
			}
		}
	}

	for i := 0; i < 1000; i++ {
		for j := 0; j < 1000; j++ {
			if matrix[i][j] >= 2 {
				ans++
			}
		}
	}

	fmt.Println("Part 2:", ans)
}

func main() {
	part1("./input.txt")
	part2("./input.txt")
}
